Introducing ClientRequestFilter.java: A New Plugin for Merging Additional Headers into Client Requests in the Authentication Filter

SthuthiGhosh9400 opened the pull request:
## Description
1. A new interface, ClientRequestFilter.java, has been introduced in the Presto SPI to customize the headers used by the Presto runtime to process queries. Some example use cases include customized authentication workflows, or enriching query attributes such as the query source. 
2. One key use case is setting extra credentials in the request headers.

## Motivation and Context
https://github.com/prestodb/presto/issues/23997

## Impact
This feature enables the merging of headers of any type, offering a versatile solution for header management.

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==
General Changes
* Add ``ClientRequestFilter.java`` interface in Presto-spi. :pr:`23380`
* Improve Request Headers in the Authentication Filter Class. :pr:`23380`

tdcmeehan reviewed (changes requested):
This PR needs some test cases.

SthuthiGhosh9400 commented:
@tdcmeehan 

I have added a test case to verify setting values in the request header and committed it.
Could you have a check and confirm ?
Thanks.

SthuthiGhosh9400 commented on presto-main/pom.xml:
```diff
@@ -499,6 +499,11 @@
             <artifactId>ratis-common</artifactId>
             <optional>true</optional>
         </dependency>
+        <dependency>
```
@tdcmeehan 
Instead of using Mockito to create mock objects and define their behavior, can I manually write simple classes (stubs) that simulate the behavior of the actual classes?

Do you have any suggestion here in order to verify the codebase if value passed to request header is there correclty?

tdcmeehan replied:
Yes, that is how we go about this.

SthuthiGhosh9400 replied:
@tdcmeehan 
I have refactored the code that was added for the test case.

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestRequestHeaderModifier.java:
```diff
@@ -0,0 +1,1211 @@
+        assertEquals("CustomValue", wrappedRequest.getHeader("X-Custom-Header"));
+    }
+
+    abstract static class HttpServletRequestAdapter
```
There's already a `MockHttpServletRequest`, can you use that?

SthuthiGhosh9400 replied:
@tdcmeehan 
I have refactored the code to use MockHttpServletRequest. Could you check it?
Thanks.

tdcmeehan commented:
While we don't use Mockito, the stubs in the test are a little out of control.  We need to reduce the size of this simple test.  I have a couple of suggestions that you can research to make the testing simpler.

* Can you look at `TestHttpRemoteTask` and examine how they implemented an in-memory endpoint for the purposes of testing?  Perhaps you can build something similar.
* `TestingPrestoServer` has all sorts of helper methods to retrieve individual classes from Presto.  Perhaps you can add a new getter to retrieve the `RequestModifierManager`, register a simple new request modifier that just adds a dummy header, and make a client call that proves the header is added.

SthuthiGhosh9400 commented:
> * `TestingPrestoServer` has all sorts of helper methods to retrieve individual classes from Presto.  Perhaps you can add a new getter to retrieve the `RequestModifierManager`, register a simple new request modifier that just adds a dummy header, and make a client call that proves the header is added.

@tdcmeehan 
I have attempted to improve the test case based on the suggestion. Could you have a check?
Thanks.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -93,9 +103,28 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-
             // authentication succeeded
-            nextFilter.doFilter(withPrincipal(request, principal), response);
+            CustomHttpServletRequestWrapper wrappedRequest = withPrincipal(request, principal);
```
Instead of modifying the authentication filter, this should just go in its own filter that is applied after the authentication filter.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -152,4 +174,65 @@ private static void skipRequestBody(HttpServletRequest request)
+            this.customHeaders = new HashMap<>();
+        }
+
+        public void addHeader(String name, String value)
```
Unused

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -152,4 +174,65 @@ private static void skipRequestBody(HttpServletRequest request)
+    public static class CustomHttpServletRequestWrapper
+            extends HttpServletRequestWrapper
+    {
+        private final Map<String, String> customHeaders;
```
Please use a concurrent map

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/RequestModifier.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Map;
+import java.util.Optional;
+
+public interface RequestModifier
```
Let's call this something more descriptive, such as `ClientRequestFilter`.

SthuthiGhosh9400 commented:
@tdcmeehan
I have incorporated the review comments. Could you please verify them?
Thanks.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -95,9 +106,28 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                    extraHeaderValueMap.ifPresent(map -> {
+                        for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                            if (request.getHeader(extraHeaderEntry.getKey()) == null) {
+                                extraHeadersMap.putIfAbsent(extraHeaderEntry.getKey(), extraHeaderEntry.getValue());
```
I don't think we should silently not put the header if there's an existing header.  I think we should consider the following:

1. The SPI contract mandates that any potential header is known upfront.
2. The runtime will validate that the headers that are requested to be modified match with what is actually returned.
3. There is a blocklist of headers that the plugin should not modify (such as query ID).
4. The runtime will validate that all registered filters only return a disjoint set of headers being added.  If there is a conflict, this is a fatal error that will require an administrator to fix (but un-registering one of these plugins).
5. If there is a conflict here, then an exception is thrown.

SthuthiGhosh9400 replied:
@tdcmeehan 
I have gone through your comments and added a few validations in the current implementation. I can verify that with you. However, I have to know about the headers in the blocklist that should not be modified.
Could you help?

tdcmeehan replied:
Let's start with the query ID, trace ID and transaction ID in PrestoHeaders.java.

SthuthiGhosh9400 replied:
I have incorporated the changes mentioned in your comment. Could you please review them and share your suggestions?

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -95,9 +107,37 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                        for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                            String headerKey = extraHeaderEntry.getKey();
+                            if (headersBlockList.contains(headerKey)) {
+                                throw new RuntimeException("Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
```
Please throw a PrestoException and introduce a new error code for this.  It should be a system error.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +206,50 @@ private static void skipRequestBody(HttpServletRequest request)
+            while (originalHeaderNames.hasMoreElements()) {
+                headerNames.add(originalHeaderNames.nextElement());
+            }
+            return Collections.enumeration(headerNames);
```
Please use Immutable collections.

tdcmeehan reviewed (changes requested):
Some feedback is still unadressed.

SthuthiGhosh9400 commented:
> Some feedback is still unadressed.

@tdcmeehan Could you help me identify which feedback is still pending? I believe I have addressed feedback items 2, 3, 4, and 5 from the list above.

SthuthiGhosh9400 commented:
> Please scroll up in this PR and examine feedback on the code that is still remaining.

Yeah. @tdcmeehan I have incorporated all feedback items now. Kindly review them and share your suggestions?

Thanks.

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,298 @@
+    private List<ClientRequestFilter> getClientRequestFilters()
+    {
+        List<ClientRequestFilter> requestFilters = new ArrayList<>();
+        ClientRequestFilter customModifier = new ClientRequestFilter()
+        {
+            @Override
+            public List<String> getHeaderNames()
+            {
+                return Collections.singletonList("X-Custom-Header");
+            }
+            @Override
+            public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+            {
+                Map<String, String> headers = new HashMap<>();
+                headers.put("X-Custom-Header", "CustomValue_1");
+                return Optional.of(headers);
+            }
+        };
+
+        ClientRequestFilter customModifierConflict = new ClientRequestFilter()
+        {
+            @Override
+            public List<String> getHeaderNames()
+            {
+                return Collections.singletonList("X-Custom-Header");
+            }
+            @Override
+            public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+            {
+                Map<String, String> headers = new HashMap<>();
+                headers.put("X-Custom-Header", "CustomValue_2");
+                return Optional.of(headers);
+            }
```
Please reduce duplicate code.  Consider an inner helper class.

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,298 @@
+    }
+
+    static class ConcreteHttpServletRequest
+            extends MockHttpServletRequest
```
Why not modify `MockHttpServletRequest` directly?

SthuthiGhosh9400 replied:
In ConcreteHttpServletRequest, I am storing additional headers in the customHeaders map and overriding the getHeaders method to manage them. This functionality isn't natively provided by MockHttpServletRequest, so extending it makes sense.

tdcmeehan replied:
Just add a new constructor to `MockHttpServletRequest` that overrides the headers.

SthuthiGhosh9400 commented:
@tdcmeehan 
I have incorporated the review comments. Kindly review them.
If it looks good, I can squash the commits and edit the commit message according to the guidelines.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -95,9 +110,37 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                        for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                            String headerKey = extraHeaderEntry.getKey();
+                            if (headersBlockList.contains(headerKey)) {
+                                throw new PrestoException(HEADER_MODIFICATION_ATTEMPT, "Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
```
Instead of mentioning a "blocked headers list", just list out the headers that are not allowed to be modified.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -95,9 +110,37 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                                throw new PrestoException(HEADER_MODIFICATION_ATTEMPT, "Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
+                            }
+                            if (globallyAddedHeaders.contains(headerKey)) {
+                                throw new RuntimeException("Header conflict detected: " + headerKey + " already added by another filter.");
```
Do not throw RuntimeException, throw PrestoException.

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,294 @@
+                    for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                        String headerKey = extraHeaderEntry.getKey();
+                        if (headersBlockList.contains(headerKey)) {
+                            throw new RuntimeException("Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
```
Why are we duplicating code here?  I am very confused about how this test works.

tdcmeehan reviewed (changes requested):
Please refactor tests to use the code under test.

SthuthiGhosh9400 commented:
@tdcmeehan 
I have refactored the test cases and used a separate method for adding headers to the request in the Authentication filter.
Could you review it and share your suggestions?

tdcmeehan commented:
Please squash commits and ensure the commit format follows our guidelines in contributing.md.

SthuthiGhosh9400 commented:
Hi @tdcmeehan 
Could you please check this pr?

tdcmeehan commented:
@SthuthiGhosh9400 the test failures look related, please look into them.

SthuthiGhosh9400 commented:
@tdcmeehan  Could you review this?

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +218,51 @@ private static void skipRequestBody(HttpServletRequest request)
+            return super.getHeaders(name);
+        }
+
+        public void setHeaders(Map<String, String> headers)
```
Can we set the headers in the constructor?

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +218,51 @@ private static void skipRequestBody(HttpServletRequest request)
         }
     }
+
+    public static class CustomHttpServletRequestWrapper
```
```suggestion
    private static class ModifiedHttpServletRequest
```

tdcmeehan replied:
All of the public methods on `ModifiedHttpServletRequest` override something from a parent interface or abstract class, so can't you use the most restrictive interface?  Wouldn't that be `HttpServletRequest`?

SthuthiGhosh9400 replied:
Hi @tdcmeehan 
When I implement the HttpServletRequest interface directly, I must implement all of its abstract methods. 
The HttpServletRequest interface contains many methods, so implementing it directly can be tedious, especially when I only need to override a few methods, such as getHeader, getHeaderNames, and getHeaders.

tdcmeehan replied:
I am not telling you to implement `HttpServletRequest`, I am simply saying that this method can just return the interface `HttpServletRequest` for better [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)).

SthuthiGhosh9400 replied:
@tdcmeehan 
I have made the change and committed. Please review the code. 
Thanks.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -126,6 +141,43 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+        Map<String, String> extraHeadersMap = new HashMap<>();
+        Set<String> globallyAddedHeaders = new HashSet<>();
+
+        for (ClientRequestFilter requestFilter : clientRequestFilterManager.getClientRequestFilters()) {
```
If the list is empty, can we short circuit and return the input?

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -126,6 +141,43 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+    public CustomHttpServletRequestWrapper mergeExtraHeaders(HttpServletRequest request, Principal principal)
+    {
+        CustomHttpServletRequestWrapper wrappedRequest = new CustomHttpServletRequestWrapper(request);
+        Map<String, String> extraHeadersMap = new HashMap<>();
+        Set<String> globallyAddedHeaders = new HashSet<>();
```
Please use immutable map builders and construct when needed.

SthuthiGhosh9400 replied:
I have used immutable map builders.

steveburnett commented:
Nit rephrasing of release note entries following the [Order of changes](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes) in the Release Notes Guidelines, and some formatting corrections from my earlier comment: 

```
== RELEASE NOTES ==
General Changes
* Add  ``RequestModifier.java`` interface in Presto-spi. :pr:`23380`
* Improve Request Headers in the Authentication Filter Class. :pr:`23380`
```

rschlussel commented:
the linked issue is an ibm internal issue. Can you share a public issue for this?

SthuthiGhosh9400 commented:
@tdcmeehan I have incorporated the review comments. Please check.
Thanks.

SthuthiGhosh9400 commented:
> the linked issue is an ibm internal issue. Can you share a public issue for this?

Thanks. I have added a public issue for this.

SthuthiGhosh9400 commented:
Hi @tdcmeehan 
Could you please review?

tdcmeehan reviewed (changes requested):
While this is sufficient for your current use case, I worry it is not generic enough for future use cases.  Let's refactor it in the following way:

1) Introduce a `ClientRequestFilterFactory`.  See by analogy `PlanCheckerProviderFactory`.  This class should take in configuration so that we can configure the filters.  The filters should also be named to differentiate them.
2) `ClientRequestFilterManager` should register the factory, not a raw client request filter.
3) There should be a method `loadClientRequestFilters` which invokes the factory calls and creates the filters.  This should populate a new immutable datastructure which can then be used to supply the filters to Jetty.

See for example: https://github.com/prestodb/presto/pull/23955

SthuthiGhosh9400 commented:
@tdcmeehan I have addressed the requested changes. Could you please review?

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +222,47 @@ private static void skipRequestBody(HttpServletRequest request)
+        public ModifiedHttpServletRequest(HttpServletRequest request, Map<String, String> headers)
+        {
+            super(request);
+            this.customHeaders = new ConcurrentHashMap<>(headers);
```
```suggestion
            this.customHeaders = new ConcurrentHashMap<>(requireNonNull(headers, "headers is null"));
```

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +222,47 @@ private static void skipRequestBody(HttpServletRequest request)
+        @Override
+        public String getHeader(String name)
+        {
+            String headerValue = customHeaders.get(name);
+            if (headerValue != null) {
+                return headerValue;
+            }
```
```suggestion
            if (customHeaders.isPresent(name)) {
                return customHeaders.get(name);
            }
```

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +222,47 @@ private static void skipRequestBody(HttpServletRequest request)
+        @Override
+        public Enumeration<String> getHeaderNames()
+        {
+            ImmutableSet.Builder<String> headerNamesBuilder = ImmutableSet.builder();
+            headerNamesBuilder.addAll(customHeaders.keySet());
+            Enumeration<String> originalHeaderNames = super.getHeaderNames();
+            while (originalHeaderNames.hasMoreElements()) {
+                headerNamesBuilder.add(originalHeaderNames.nextElement());
+            }
+            return Collections.enumeration(headerNamesBuilder.build());
```
```suggestion
            return enumeration(ImmutableSet.builder()
                    .addAll(customHeaders.keySet())
                    .addAll(list(super.getHeaderNames()))
                    .build());
```

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +222,47 @@ private static void skipRequestBody(HttpServletRequest request)
+        public Enumeration<String> getHeaders(String name)
+        {
+            if (customHeaders.containsKey(name)) {
+                return Collections.enumeration(Collections.singleton(customHeaders.get(name)));
```
```suggestion
                return enumeration(immutableListOf(customHeaders.get(name)));
```

Please static import as described in our contributing guide.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/PrestoServer.java:
```diff
@@ -138,7 +140,8 @@ public void run()
                 new NodeTtlFetcherManagerModule(),
-                new ClusterTtlProviderManagerModule());
+                new ClusterTtlProviderManagerModule(),
+                new ClientRequestFilterModule());
```
Just add this inline to `HttpServerModule`, it's a one liner and strongly aligned with HTTP.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -861,4 +865,11 @@ private static int driftServerPort(DriftServer server)
+
+    public static ClientRequestFilterManager getClientRequestFilterManager(List<ClientRequestFilterFactory> requestFilterFactory)
+    {
+        ClientRequestFilterManager manager = new ClientRequestFilterManager();
```
Shouldn't this be retrieved from Guice, since you're already creating one in `ClientRequestServerModule`?

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/ClientRequestFilter.java:
```diff
@@ -0,0 +1,25 @@
+{
+    List<String> getHeaderNames();
+
+    <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo);
```
Why are we parameterizing `additionalInfo`?

SthuthiGhosh9400 replied:
Since additionalInfo can vary in type, parameterizing it (<T>) is a good idea because it offers flexibility and reusability.

tdcmeehan replied:
Usually, plugins add a context object, and the runtime will populate the context object.

In this case, create a new `ClientRequestFilterContext`.  At first it has just one field, `Principal`, but it can be extended in the future.

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/Plugin.java:
```diff
@@ -142,4 +143,14 @@ default Iterable<NodeStatusNotificationProviderFactory> getNodeStatusNotificatio
+        return emptyList();
+    }
+
+    default Iterable<PlanCheckerProviderFactory> getPlanCheckerProviderFactories()
+    {
+        return emptyList();
+    }
```
Why add this?

SthuthiGhosh9400 commented:
@tdcmeehan I have addressed the comments. Could you please take a look?

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,193 @@
+                @Override
+                public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+                {
+                    Map<String, String> headers = new HashMap<>();
+                    headers.put(headerName, headerValue);
+                    return Optional.of(headers);
```
```suggestion
                    return Optional.of(ImmutableMap.of(headerName, headerValue));
```

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,193 @@
+
+    private List<ClientRequestFilterFactory> createFilterFactories(String[][] filterConfigs)
+    {
+        List<ClientRequestFilterFactory> factories = new ArrayList<>();
+        for (String[] config : filterConfigs) {
+            factories.add(new GenericClientRequestFilterFactory(config[0], config[1], config[2]));
+        }
+        return factories;
```
```suggestion
        ImmutableList.Builder<ClientRequestFilterFactory> factories = ImmutableList.builder();
        for (String[] config : filterConfigs) {
            factories.add(new GenericClientRequestFilterFactory(config[0], config[1], config[2]));
        }
        return factories.build();
```

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/server/MockHttpServletRequest.java:
```diff
@@ -56,6 +56,12 @@ public MockHttpServletRequest(ListMultimap<String, String> headers, String remot
         this.attributes = new HashMap<>(requireNonNull(attributes, "attributes is null"));
     }
 
+    public MockHttpServletRequest(ListMultimap<String, String> headers)
+    {
+        // Default remoteAddress and empty attributes
+        this(headers, "127.0.0.1", new HashMap<>());
+    }
```
```suggestion
    public MockHttpServletRequest(ListMultimap<String, String> headers)
    {
        // Default remoteAddress and empty attributes
        this(headers, DEFAULT_ADDRESS, ImmutableMap.of());
    }
```

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/ClientRequestFilterManager.java:
```diff
@@ -0,0 +1,61 @@
+    private List<ClientRequestFilterFactory> factories = new CopyOnWriteArrayList<>();
+    @GuardedBy("this")
+    private volatile List<ClientRequestFilter> filters = ImmutableList.of();
+    AtomicBoolean loaded = new AtomicBoolean();
```
```suggestion
    private final AtomicBoolean loaded = new AtomicBoolean();
```

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -166,4 +227,43 @@ private static void skipRequestBody(HttpServletRequest request)
+        public ModifiedHttpServletRequest(HttpServletRequest request, Map<String, String> headers)
+        {
+            super(request);
+            this.customHeaders = new ConcurrentHashMap<>(requireNonNull(headers, "headers is null"));
```
```suggestion
            this.customHeaders = ImmutableMap.of(requireNonNull(headers, "headers is null"));
```

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,192 @@
+
+public class TestClientRequestFilterPlugin
+{
+    static class GenericClientRequestFilterFactory
```
Move test classes to the bottom.  Put fields of `TestClientRequestFilterPlugin` first, then tests, then helper methods, then inner classes.

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/ClientRequestFilterContext.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.security.Principal;
+
+public class ClientRequestFilterContext
+{
+    private Principal principal;
+
+    public ClientRequestFilterContext(Principal principal)
+    {
+        this.principal = principal;
+    }
+
+    public Principal getPrincipal()
+    {
+        return principal;
+    }
+
+    public void setPrincipal(Principal principal)
+    {
+        this.principal = principal;
+    }
+}
```
```suggestion
public class ClientRequestFilterContext
{
    private final Principal principal;

    public ClientRequestFilterContext(Principal principal)
    {
        this.principal = principal;
    }

    public Principal getPrincipal()
    {
        return principal;
    }
}
```

1. The setter is not used.
2. As described in our developer guide, make it immutable.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -126,6 +142,50 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                    .allMatch(headerName -> request.getHeader(headerName) != null);
+
+            if (!headersPresent) {
+                ClientRequestFilterContext context = new ClientRequestFilterContext(principal);
+                Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(context);
```
Sorry, I realize now that the principal is per request.  Context objects typically are used to create the plugin.  So please revert this, and just add the principal as a plain old parameter to `getExtraHeaders`, no need for fancy parameterization.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java:
```diff
@@ -126,6 +142,50 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+            }
+        }
+
+        Map<String, String> extraHeadersMap = extraHeadersMapBuilder.build();
+        return new ModifiedHttpServletRequest(request, extraHeadersMapBuilder.build());
```
```suggestion
        return new ModifiedHttpServletRequest(request, extraHeadersMapBuilder.build());
```

Remove duplicate build.

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java:
```diff
@@ -0,0 +1,192 @@
+        @Override
+        public ClientRequestFilter create(String filterName)
+        {
+            return new ClientRequestFilter() {
+                @Override
+                public List<String> getHeaderNames()
+                {
+                    return ImmutableList.of(headerName);
+                }
+
+                @Override
+                public Optional<Map<String, String>> getExtraHeaders(ClientRequestFilterContext context)
+                {
+                    return Optional.of(ImmutableMap.of(headerName, headerValue));
+                }
+            };
```
Make this an inner class.

tdcmeehan reviewed (commented):
Almost there.  Please also add documentation for this plugin to https://github.com/prestodb/presto/tree/master/presto-docs/src/main/sphinx/develop

tdcmeehan commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,23 @@
+
+After installing a plugin that implements ``ClientRequestFilterFactory`` on the coordinator, the ``AuthenticationFilter`` class passes the ``principal`` object to the request filter, which returns the header values as a map.
+
+Presto uses the name of the request filter to determine whether the header is present in the blocklist.
```
What blocklist?  Mention the headers which are not allowed to be overidden.  Link to the Java source of headers that may be overidden.

SthuthiGhosh9400 replied:
Added.

SthuthiGhosh9400 commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,23 @@
+======================
+
+Presto facilitates seamless integration of additional headers into client requests using the 
+Client Request Filter plugin via the AuthenticationFilter class, enabling efficient and dynamic header management.
```
Sure.

SthuthiGhosh9400 commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,23 @@
+--------------
+
+The ``ClientRequestFilterFactory`` is responsible for creating instances of ``ClientRequestFilter``. It also defines 
+the name of the filter, which is subsequently used as the header name in client requests.
```
Corrected.

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/ClientRequestFilter.java:
```diff
@@ -0,0 +1,26 @@
+
+public interface ClientRequestFilter
+{
+    List<String> getHeaderNames();
```
```suggestion
    Set<String> getExtraHeaderKeys();
```

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/ClientRequestFilter.java:
```diff
@@ -0,0 +1,26 @@
+{
+    List<String> getHeaderNames();
+
+    Optional<Map<String, String>> getExtraHeaders(Principal principal);
```
Can it just return empty map and we don't return optional?

tdcmeehan commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,23 @@
+The ``ClientRequestFilterFactory`` is responsible for creating instances of ``ClientRequestFilter``. It also defines 
+the name of the filter.
+
+``ClientRequestFilter``  interface provides two methods: ``getExtraHeaders()``,
+which returns header values as key-value pairs in a map, and ``getHeaderNames()``, which returns a list of header names used as the header names in client requests.
```
I imagine `getExtraHeaders` is used for the runtime to quickly check if it needs to apply a more expensive call to enrich the headers.  If so, add this context to the documentation.

SthuthiGhosh9400 replied:
Added.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/ClientRequestFilterManager.java:
```diff
@@ -0,0 +1,61 @@
+@ThreadSafe
+public class ClientRequestFilterManager
+{
+    private List<ClientRequestFilterFactory> factories = new CopyOnWriteArrayList<>();
```
Let's make this a map.

1) The key of the map should be the name of the factory.
2) We should not allow duplicate names.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/ClientRequestFilterManager.java:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        filters = factories.stream()
+                .map(factory -> factory.create(factory.getName()))
```
This construct is strange.  See above comment.

tdcmeehan replied:
What I meant was, please don't pass in `factory.getName()` into the `create` method, there is no need because it can just internally call `getName()`.

tdcmeehan reviewed (changes requested):
Just one more thing, otherwise looks good.

SthuthiGhosh9400 commented:
@tdcmeehan I have made the changes. Could you please take a look?

steveburnett commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,22 @@
+Client Request Filter
+======================
+
+Presto allows operators to customize the headers used by the Presto runtime to process queries. Some example use cases include customized authentication workflows, or enriching query attributes such as the query source. This can be achieved using the Client Request Filter plugin, which allows control over header customization during query execution.
```
```suggestion
Presto allows operators to customize the headers used to process queries. Some example use cases include customized authentication workflows, or enriching query attributes such as the query source. Use the Client Request Filter plugin to control header customization during query execution.
```
Some suggestions for simpler phrasing and conciseness.

steveburnett commented on presto-docs/src/main/sphinx/develop/client-request-filter.rst:
```diff
@@ -0,0 +1,22 @@
+
+After installing a plugin that implements ``ClientRequestFilterFactory`` on the coordinator, the ``AuthenticationFilter`` class passes the ``principal`` object to the request filter, which returns the header values as a map.
+
+Presto uses the request filter to determine whether a header is present in the blocklist. The blocklist includes headers such as ``X-Presto-Transaction-Id``, ``X-Presto-Started-Transaction-Id``, ``X-Presto-Clear-Transaction-Id``, and ``X-Presto-Trace-Token``, which are not allowed to be overridden. For a complete list of headers that may be overridden, please refer to the `Java source`_ (note tha ...
```
```suggestion
Presto uses the request filter to determine whether a header is present in the blocklist. The blocklist includes headers such as ``X-Presto-Transaction-Id``, ``X-Presto-Started-Transaction-Id``, ``X-Presto-Clear-Transaction-Id``, and ``X-Presto-Trace-Token``, which are not allowed to be overridden. 

For a complete list of headers, see the `Java source`_. 

Note: The `Java source`_ includes these blocklist headers that are not eligible for overriding. The other headers not mentioned here can be overridden.
```
Some rephrasing, and adding line breaks for ease of reading. 

I made the assumption in my draft suggestion here that the headers you list are the complete list of the headers that cannot be overridden, and that the other headers in the Java source can all be overridden. If my assumption is wrong, please revise as you feel correct.

steveburnett reviewed (changes requested):
Thanks for the doc! A few suggestions. 

Also, to have this new doc page be visible in the [Developer Guide index page](https://prestodb.io/docs/current/develop.html), you must edit [presto-docs/src/main/sphinx/develop.rst](https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/develop.rst) 
and add 
`develop/client-request-filter`.

SthuthiGhosh9400 commented:
@steveburnett I have incorporated the changes requested. Thanks.

steveburnett reviewed (dismissed):
LGTM! (docs)

Pull updated branch, new local doc build, looks good. Thanks!

SthuthiGhosh9400 commented:
@tdcmeehan Could you take a look and approve?

SthuthiGhosh9400 commented:
Hi @tdcmeehan 
I have resolved the conflicts with rebase. Could you help?

SthuthiGhosh9400 commented:
If it looks good, can we merge this PR? @tdcmeehan

tdcmeehan commented:
@SthuthiGhosh9400 the CI is read.  Please investigate.  If you find it's not related, rebase to rerun the test.

SthuthiGhosh9400 commented:
@tdcmeehan I reran the test, and it looks good now.

SthuthiGhosh9400 commented:
@tdcmeehan Could we proceed with merging this PR?
